components/brave_ads/browser/ads_notification_handler.h

```diff
@@ -0,0 +1,116 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_ADS_BROWSER_ADS_NOTIFICATION_HANDLER_H_
+#define BRAVE_COMPONENTS_BRAVE_ADS_BROWSER_ADS_NOTIFICATION_HANDLER_H_
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "base/callback.h"
+#include "base/memory/weak_ptr.h"
+#include "base/supports_user_data.h"
+#include "chrome/browser/notifications/notification_handler.h"
+#include "url/gurl.h"
+
+namespace content {
+class BrowserContext;
+}  // namespace content
+
+namespace brave_ads {
+
+class AdsServiceImpl;
+
+const void* const kAdsNotificationHandlerKey = &kAdsNotificationHandlerKey;
+
+class AdsNotificationHandler : public NotificationHandler {
+ public:
+  explicit AdsNotificationHandler(content::BrowserContext* browser_context);
+  ~AdsNotificationHandler() override;
+
+  // NotificationHandler implementation
+  void OnShow(
+      Profile* profile,
+      const std::string& id) override;
+  void OnClose(
+      Profile* profile,
+      const GURL& origin,
+      const std::string& id,
+      bool by_user,
+      base::OnceClosure completed_closure) override;
+  void OnClick(
+      Profile* profile,
+      const GURL& origin,
+      const std::string& id,
+      const base::Optional<int>& action_index,
+      const base::Optional<base::string16>& reply,
+      base::OnceClosure completed_closure) override;
+  void DisableNotifications(
+      Profile* profile,
+      const GURL& origin) override;
+  void OpenSettings(
+      Profile* profile,
+      const GURL& origin) override;
+
+  void SetAdsService(brave_ads::AdsServiceImpl* ads_service);
+
+  static const void* UserDataKey();
+
+  // This class is used to set the AdsNotificationHandler as user data
+  // on a BrowserContext. This is used instead of AdsNotificationHandler
+  // directly because SetUserData requires a std::unique_ptr. This is
+  // safe because we remove the user data in AdsNotificationHandler's
+  // destructor.
+  class UnownedPointer : public base::SupportsUserData::Data {
+   public:
+    explicit UnownedPointer(AdsNotificationHandler* pointer)
+        : pointer_(pointer) {}
+
+    AdsNotificationHandler* get() const { return pointer_; }
+
+   private:
+    AdsNotificationHandler* const pointer_;
+
+    DISALLOW_COPY_AND_ASSIGN(UnownedPointer);
+  };
+
+ private:
+  struct ShowNotification {
+    ShowNotification(const std::string& id, Profile* profile)
+        : id_(id), profile_(profile) {}
+
+    std::string id_;
+    Profile* profile_;
+  };
+
+  struct CloseNotification {
+    CloseNotification(const std::string& id,
+                      Profile* profile,
+                      const GURL& origin,
+                      bool by_user);
+    ~CloseNotification();
+
+    std::string id_;
+    Profile* profile_;
+    GURL origin_;
+    bool by_user_;
+  };
+
+  void SendCachedNotifications();
```

Comment:
cached is actually probably not the right word to use, I think `pending` makes more sense here